File: Greedy/FractionalKnapsack/fractionalKnapsack.js
(function(){



    function fKnapsack(arr,n,W){
        let currentWeight = 0;
        let fValue = 0;

        for (let i of arr) {
            if (currentWeight + i.weight <= W) {
                currentWeight += i.weight;
                fValue += i.value;
            } else {
                let remainingWeight = W - currentWeight;
                fValue += (remainingWeight / i.weight) * i.value;
                break;
            }
        }
        return fValue;
    }


    (function main(){


        let value = [70,80,90];
        let weight = [15,20,40];
        let maxWeight = 70;

        let Item = [];
        for (let i=0; i<value.length; i++) {
            let item = {};
            item.value = value[i];
            item.weight = weight[i];
            Item.push(item);
        }

        // Sort array
        Item.sort((a,b) => {
            let value1 = parseFloat(a.value/a.weight);
            let value2 = parseFloat(b.value / b.weight);
            if (value1 > value2) {
                return 1;
            }
        });

        console.log(fKnapsack(Item,Item.length,maxWeight));
        

    }());


}());

File: Greedy/FractionalKnapsack/fractionalKnapsack.cpp
#include<iostream>
#include<algorithm>
using std::cin;
using std::cout;
using std::endl;
using std::sort;

struct Item {
    int value;
    int weight;
};

bool compare(Item a, Item b) {
    double value1 {((double)a.value) / (a.weight)};
    double value2 {((double)b.value) / (b.weight)};
    return value1 > value2;
}

double fKnapsack(Item* arr, int n, int W) {
    sort(arr, arr + n, compare);
    int currentWeight {};
    double fValue {};

    for(int i{}; i < n; i++) {
        if (currentWeight + arr[i].weight <= W) {
            currentWeight += arr[i].weight;
            fValue += arr[i].value;
        } else {
            int remainingWeight {W - currentWeight};
            fValue += arr[i].value * ((double)remainingWeight / arr[i].weight);
            break;
        }
    }
    return fValue;
}

int main() {
    int n;
    cin >> n;
    Item arr[n];
    for(int i{}; i < n; i++) cin >> arr[i].value >> arr[i].weight;

    int W;
    cin >> W;
    double ans {fKnapsack(arr, n, W)};
    cout << ans << endl;
}
